Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue11853.xaml.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Text;
+using Xamarin.Forms.CustomAttributes;
+using Xamarin.Forms.Internals;
+using System.Threading.Tasks;
+using Xamarin.Forms.Xaml;
+using System.ComponentModel;
+using System.Linq;
+using System.Windows.Input;
+
+#if UITEST
+using Xamarin.UITest;
+using NUnit.Framework;
+using Xamarin.Forms.Core.UITests;
+#endif
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	[Issue(IssueTracker.Github, 11853, "[Bug][iOS] Concurrent issue leading to crash in SemaphoreSlim.Release in ObservableItemsSource",
+		PlatformAffected.iOS)]
+#if UITEST
+	[NUnit.Framework.Category(UITestCategories.CollectionView)]
+#endif
+	public partial class Issue11853 : TestContentPage
+	{
+		const string Run = "Run";
+
+		protected override void Init() { }
+
+		public Issue11853()
+		{
+#if APP
+			InitializeComponent();
+#endif
+			BindingContext = new _11853ViewModel();
+		}
+
+#if UITEST
+		[Test]
+		public void JustWhalingAwayOnTheCollectionViewWithAddsAndClearsShouldNotCrash()
+		{
+			RunningApp.WaitForElement(Run);
+			RunningApp.Tap(Run);
+			Task.Delay(5000).Wait();
+			RunningApp.Tap(Run);
+			Task.Delay(5000).Wait();
+
+			// If we can still find the button, then we didn't crash
+			RunningApp.WaitForElement(Run);
+		}
+#endif
+
+		public class _11853Item
+		{
+			public string Text { get; set; }
+		}
+
+		public class _11853ViewModel : INotifyPropertyChanged
+		{
+			private bool isListVisible;
+			public event PropertyChangedEventHandler PropertyChanged;
+
+			public bool IsListVisible
+			{
+				get => isListVisible;
+				set
+				{
+					isListVisible = value;
+					PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsListVisible)));
+				}
+			}
+
+			public ObservableCollection<_11853Item> Items { get; }
+			public ICommand TestCommand { get; }
+
+			public _11853ViewModel()
+			{
+				Items = new ObservableCollection<_11853Item>();
+				TestCommand = new Command(async () =>
+				{
+					var items = CreateItems(10, 0).ToList().First();
+					var items2 = CreateItems(10, 0).ToList().Skip(1).First();
+
+					int iterations = 1000; //10000;
+
+					for (var i = 0; i < iterations; i++)
+					{
+						await Task.Delay(1);
```

Comment:
The example is great, but to cover all cases, could we add to the example?
- Clear all the items and add again.
- Swap the collection.

Maybe add another two buttons for these cases.

Reply:
What do you mean by "swap the collection"?

Reply:
> Clear all the items and add again.

That's what it already does - it adds everything, clears it, and adds again in a loop. 1000 times.

Reply:
With swap, I mean create a new `ObservableCollection` and assign to Items.